File: frontend/src/types/quiz.ts
/**
 * Quiz types — mirror backend Pydantic schemas in
 * backend/app/schemas/quiz.py
 */

export type Difficulty = 'easy' | 'medium' | 'hard'
export type QuestionType = 'mcq' | 'short_answer'

export interface QuestionStudentView {
  id: string
  order_index: number
  question_text: string
  question_type: QuestionType
  options: string[] | null
  difficulty: Difficulty
  topic: string | null
}

export interface QuestionTeacherView extends QuestionStudentView {
  correct_answer: string
  explanation: string | null
}

export interface QuizSummary {
  id: string
  subject_id: string
  document_id: string | null
  created_by_id: string
  title: string
  description: string | null
  difficulty: Difficulty
  time_limit_minutes: number | null
  /** Exact per-attempt duration in seconds. Source of truth — minutes is kept
   *  for legacy display only. */
  time_limit_seconds: number | null
  /** ISO 8601 UTC timestamps for the open / close window. null = no bound. */
  opens_at: string | null
  closes_at: string | null
  is_published: boolean
  is_ai_generated: boolean
  created_at: string
  question_count: number
  attempt_count: number
  avg_score: number | null
  subject_code: string | null
  subject_name: string | null
  /** Set on student-facing responses only — true iff this student has already
   *  submitted an attempt. Drives the "Take" → "View Result" UI switch. */
  has_attempted: boolean | null
}

export interface QuizForStudent extends QuizSummary {
  questions: QuestionStudentView[]
}

export interface QuizForTeacher extends QuizSummary {
  questions: QuestionTeacherView[]
}

export interface QuizGenerateRequest {
  subject_id: string
  /** Legacy single doc — backend folds this into document_ids if both given. */
  document_id?: string | null
  /** Multi-doc selection. null/empty/undefined = backend uses all docs in
   *  the subject. Non-empty = use only those documents. */
  document_ids?: string[] | null
  topic_hint?: string | null
  num_questions?: number
  difficulty?: Difficulty
}

export interface QuestionUpdate {
  id?: string | null
  question_text: string
  question_type?: QuestionType
  options?: string[] | null
  correct_answer: string
  explanation?: string | null
  difficulty?: Difficulty
  topic?: string | null
  order_index?: number
}

export interface QuizUpdate {
  title?: string
  description?: string
  difficulty?: Difficulty
  time_limit_minutes?: number
  time_limit_seconds?: number
  /** Send ISO strings (use new Date(...).toISOString()) — backend stores UTC. */
  opens_at?: string | null
  closes_at?: string | null
  is_published?: boolean
  questions?: QuestionUpdate[]
}

export interface QuestionAnswer {
  question_id: string
  student_answer: string
}

/** Student-raised flag on a question ("this seems broken / ambiguous").
 *  One row per (student, question) — re-flagging just updates the reason. */
export interface QuestionFlag {
  id: string
  question_id: string
  student_id: string
  reason: string | null
  created_at: string
  updated_at: string
}

export interface QuestionFlagCreate {
  reason?: string | null
}

export interface QuizAttemptCreate {
  answers: QuestionAnswer[]
  time_taken_seconds?: number | null
}

export interface GradedAnswer {
  question_id: string
  question_text: string
  student_answer: string
  correct_answer: string
  is_correct: boolean
  topic: string | null
  difficulty: Difficulty
  explanation: string | null
}

export interface QuizAttemptResponse {
  id: string
  quiz_id: string
  student_id: string
  score: number
  total_questions: number
  correct_count: number
  time_taken_seconds: number | null
  completed_at: string
  graded_answers: GradedAnswer[]
  weak_topics: string[]
  next_difficulty_recommendation: Difficulty | null
}

export interface AttemptHistoryRow {
  id: string
  quiz_id: string
  quiz_title: string
  subject_code: string
  subject_name: string
  score: number
  total_questions: number
  correct_count: number
  time_taken_seconds: number | null
  difficulty: Difficulty
  completed_at: string
}

export interface WeakAreaResponse {
  topic: string
  subject_code: string | null
  subject_name: string | null
  score_percent: number
  attempts_count: number
  suggestion: string
}

File: backend/app/schemas/quiz.py
"""Pydantic schemas for the Quiz Engine (Phase 11, F3)."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

DifficultyLiteral = Literal["easy", "medium", "hard"]
QuestionTypeLiteral = Literal["mcq", "short_answer"]


# ── Question schemas ──

class QuestionBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_index: int
    question_text: str
    question_type: QuestionTypeLiteral
    options: list[str] | None = None
    difficulty: DifficultyLiteral
    topic: str | None = None


class QuestionStudentView(QuestionBase):
    """What students see while taking a quiz — answers and explanations are hidden."""

    pass


class QuestionTeacherView(QuestionBase):
    """Full question record for teachers and post-submission review."""

    correct_answer: str
    explanation: str | None = None


# ── Quiz schemas ──

class QuizBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    subject_id: uuid.UUID
    document_id: uuid.UUID | None
    created_by_id: uuid.UUID
    title: str
    description: str | None = None
    difficulty: DifficultyLiteral
    time_limit_minutes: int | None = None
    # Per-attempt duration as exact seconds (e.g. 5400 = 1h 30min). Frontend
    # splits this into h/m/s when rendering the countdown.
    time_limit_seconds: int | None = None
    # The window during which students may take this quiz. NULL = unbounded.
    opens_at: datetime | None = None
    closes_at: datetime | None = None
    is_published: bool
    is_ai_generated: bool
    created_at: datetime
    question_count: int = 0
    attempt_count: int = 0
    avg_score: float | None = None
    # Whether the CURRENT user has already submitted an attempt. Only
    # populated for student-facing responses; teachers get None.
    has_attempted: bool | None = None


class QuizResponse(QuizBase):
    """Quiz summary used in list views."""

    subject_code: str | None = None
    subject_name: str | None = None


class QuizForStudent(QuizResponse):
    """Quiz returned to students — questions don't include correct answers."""

    questions: list[QuestionStudentView] = []


class QuizForTeacher(QuizResponse):
    """Full quiz returned to teachers / admins (with answers + explanations)."""

    questions: list[QuestionTeacherView] = []


# ── Quiz generation ──

class QuizGenerateRequest(BaseModel):
    subject_id: uuid.UUID
    # OLD: document_id (single). Kept for back-compat if any callers still
    # send it — the service folds it into document_ids before processing.
    document_id: uuid.UUID | None = None
    # NEW: list of document IDs to draw chunks from.
    #   - None or empty list  → use ALL documents in the subject
    #   - non-empty list      → only those documents
    document_ids: list[uuid.UUID] | None = None
    topic_hint: str | None = Field(None, max_length=255)
    num_questions: int = Field(5, ge=3, le=20)
    difficulty: DifficultyLiteral = "medium"


# ── Quiz updates ──

class QuestionUpdate(BaseModel):
    id: uuid.UUID | None = None  # None = new question
    question_text: str
    question_type: QuestionTypeLiteral = "mcq"
    options: list[str] | None = None
    correct_answer: str
    explanation: str | None = None
    difficulty: DifficultyLiteral = "medium"
    topic: str | None = None
    order_index: int = 0


class QuizUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    difficulty: DifficultyLiteral | None = None
    time_limit_minutes: int | None = None
    # Either time_limit_seconds OR time_limit_minutes can be sent. seconds
    # wins if both are present (more precise).
    time_limit_seconds: int | None = None
    # Pass null to clear the window (back to always-open). Datetimes must be
    # timezone-aware; the frontend serializes JS Date with .toISOString().
    opens_at: datetime | None = None
    closes_at: datetime | None = None
    is_published: bool | None = None
    questions: list[QuestionUpdate] | None = None


# ── Attempts ──

class QuestionAnswer(BaseModel):
    question_id: uuid.UUID
    student_answer: str


# ── Question flags ──

class QuestionFlagCreate(BaseModel):
    """Body of POST .../flag. `reason` is optional — students sometimes flag
    without typing anything, just to mark the question."""

    reason: str | None = Field(None, max_length=1000)


class QuestionFlagResponse(BaseModel):
    """Read shape returned to students + teachers."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    question_id: uuid.UUID
    student_id: uuid.UUID
    reason: str | None = None
    created_at: datetime
    updated_at: datetime


class QuizAttemptCreate(BaseModel):
    answers: list[QuestionAnswer] = Field(..., min_length=1)
    time_taken_seconds: int | None = None


class GradedAnswer(BaseModel):
    question_id: uuid.UUID
    question_text: str
    student_answer: str
    correct_answer: str
    is_correct: bool
    topic: str | None = None
    difficulty: DifficultyLiteral
    explanation: str | None = None


class QuizAttemptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quiz_id: uuid.UUID
    student_id: uuid.UUID
    score: float
    total_questions: int
    correct_count: int
    time_taken_seconds: int | None = None
    completed_at: datetime
    graded_answers: list[GradedAnswer] = []
    weak_topics: list[str] = []
    next_difficulty_recommendation: DifficultyLiteral | None = None


class AttemptHistoryRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quiz_id: uuid.UUID
    quiz_title: str
    subject_code: str
    subject_name: str
    score: float
    total_questions: int
    correct_count: int
    time_taken_seconds: int | None
    difficulty: DifficultyLiteral
    completed_at: datetime


class WeakAreaResponse(BaseModel):
    topic: str
    subject_code: str | None = None
    subject_name: str | None = None
    score_percent: float
    attempts_count: int
    suggestion: str
